finality-aleph/src/sync/handler.rs

```diff
@@ -307,21 +302,111 @@ where
                 }
             }
         }
-        Ok(SyncAction::request_response(justifications))
+        Ok(Some(NetworkData::RequestResponse(
+            justifications,
+            Vec::new(),
+            Vec::new(),
+        )))
     }
 
-    /// Handle a single justification.
+    /// Handle a single unverified justification.
     /// Return `Some(id)` if this justification was higher than the previously known highest justification.
-    pub fn handle_justification(
+    fn handle_justification(
         &mut self,
         justification: J::Unverified,
-        peer: Option<I>,
+        maybe_peer: Option<I>,
     ) -> Result<Option<BlockIdFor<J>>, <Self as HandlerTypes>::Error> {
         let justification = self
             .verifier
             .verify(justification)
             .map_err(Error::Verifier)?;
-        self.handle_verified_justification(justification, peer)
+        let id = justification.header().id();
+        let maybe_id = match self
+            .forest
+            .update_justification(justification, maybe_peer)?
+        {
+            true => Some(id),
+            false => None,
+        };
+        self.try_finalize()?;
+        Ok(maybe_id)
+    }
+
+    fn handle_justifications(
+        &mut self,
+        justifications: Vec<J::Unverified>,
+        maybe_peer: Option<I>,
+    ) -> (Option<BlockIdFor<J>>, Option<<Self as HandlerTypes>::Error>) {
```

Comment:
here also we could return boolean. 
Maybe_id could be an id of the top_justification.  In the loop in the ok variant we could simply check if the justification.id.number Is higher than new_top_justification. wdyt?

Reply:
That doesn't work in case we encounter an invalid justification halfway through – then we would either return `true` (making us request something that might not make sense) or `false` (making us not request out top justification).

Reply:
let's leave this for later, when we reexamine the task system (requesting highest justified blocks)